src/twisted/conch/ssh/transport.py

```diff
@@ -637,83 +659,186 @@ def sendPacket(self, messageType, payload):
-        packet = (
-            struct.pack("!LB", totalSize + lenPad - 4, lenPad)
-            + payload
-            + randbytes.secureRandom(lenPad)
+
+        # Create the unencrypted packet.
+        payloadPacket = (
+            struct.pack("!B", lenPad) + payload + randbytes.secureRandom(lenPad)
+        )
+
+        if isETM:
+            # For Encrypt-then-MAC the total size is not encrypted.
+            encryptedPayload = self.currentEncryptions.encrypt(payloadPacket)
+            totalSizePacket = struct.pack("!L", len(encryptedPayload))
+            macTarget = totalSizePacket + encryptedPayload
+            macPacket = self.currentEncryptions.makeMAC(
+                self.outgoingPacketSequence, macTarget
+            )
+            return macTarget + macPacket
+
+        # For non ETM, we subtract the 4 bytes added to compute the
+        # padding size.
+        # But we also encrypt the package length.
+        totalSizePacket = struct.pack("!L", payloadSize + lenPad - 4)
+        encryptedPacket = self.currentEncryptions.encrypt(
+            totalSizePacket + payloadPacket
+        )
+        macTarget = totalSizePacket + payloadPacket
+
+        macPacket = self.currentEncryptions.makeMAC(
+            self.outgoingPacketSequence, macTarget
         )
-        encPacket = self.currentEncryptions.encrypt(
-            packet
-        ) + self.currentEncryptions.makeMAC(self.outgoingPacketSequence, packet)
-        self.transport.write(encPacket)
-        self.outgoingPacketSequence += 1
 
-    def getPacket(self):
+        return encryptedPacket + macPacket
+
+    def getPacket(self) -> bytes | None:
         """
         Try to return a decrypted, authenticated, and decompressed packet
         out of the buffer.  If there is not enough data, return None.
 
-        @rtype: L{str} or L{None}
-        @return: The decoded packet, if any.
+        @return: The decoded packet, if any. C{None} is returned if no packet was received yet.
         """
-        bs = self.currentEncryptions.decBlockSize
-        ms = self.currentEncryptions.verifyDigestSize
-        if len(self.buf) < bs:
+        if len(self.buf) < self.currentEncryptions.decBlockSize:
             # Not enough data for a block
             return
+
+        try:
+            if self.currentEncryptions.isInMACETM():
+                payload = self._getPacketETM()
+            else:
+                payload = self._getPacketMTE()
+
+        except _InvalidPacket as error:
+            self.sendDisconnect(
+                DISCONNECT_PROTOCOL_ERROR,
+                error.args[0].encode("ascii"),
+            )
+            return
+
+        if payload is None:
+            # No packet received yet.
+            return
+
+        if self.incomingCompression:
+            try:
+                payload = self.incomingCompression.decompress(payload)
+            except Exception:
+                self._log.failure("Error decompressing payload")
```

Comment:
love to see the new logging system used as intended :)

Reply:
As far as I know, this code is already in trunk.


I don't have much exprience with Twisted log system.

I remember that at some point I tried to document the new logging system, how to use and how to write test for it, but it wasn't a succesfull PR.